File: LeetCode/2300/Solution.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int n = spells.length;
        int[] result = new int[n];

        Arrays.sort(potions);

        for(int i=0;i<n;++i) {
            int start = 0;
            int end = potions.length-1;
            while(start <= end) {
                int mid = (start+end)/2;
                if(potions[mid]*(long)spells[i] < success) {
                    start = mid+1;
                } else {
                    end = mid-1;
                }
            }
            result[i] = potions.length-start;
        }
        return result;
    }
}

File: LeetCode/2300/Solution.kt
class Solution {
    fun successfulPairs(spells: IntArray, potions: IntArray, success: Long): IntArray {
        val n = spells.size
        val m = potions.size

        potions.sort()

        val ret = IntArray(n)
        for((i, v) in spells.withIndex()) {
            var left = 0
            var right = m-1
            while(left <= right) {
                var mid: Int = (left + right) / 2
                if(potions[mid] * v.toLong() >= success) {
                    right = mid-1
                } else {
                    left = mid + 1
                }
            }
            ret[i] = m - left
        }
        return ret
    }
}
